src/serializer/functions.js

```diff
@@ -27,22 +28,38 @@ export class Functions {
     this.functions = functions;
     this.moduleTracer = moduleTracer;
     this.writeEffects = new Map();
-    this.nameToFunctionValue = new Map();
+    this.functionExpressions = new Map();
   }
 
   realm: Realm;
   functions: ?Array<string>;
+  // maps back from FunctionValue to the expression string
+  functionExpressions: Map<FunctionValue, string>;
   moduleTracer: ModuleTracer;
-  writeEffects: Map<string, Effects>;
-  nameToFunctionValue: Map<string, FunctionValue>;
+  writeEffects: Map<FunctionValue, Effects>;
 
   checkThatFunctionsAreIndependent() {
     let functions = this.functions;
-    invariant(functions, "This method should only be called if initialized with defined functions");
+    let recordedAdditionalFunctions: Map<FunctionValue, string> = new Map();
+    let realm = this.realm;
+    let globalRecordedAdditionalFunctionsMap = this.moduleTracer.modules.logger.tryQuery(
+      () => Get(realm, realm.$GlobalObject, "__additionalFunctions"),
+      realm.intrinsics.undefined,
+      false
+    );
+    invariant(globalRecordedAdditionalFunctionsMap instanceof ObjectValue);
+    for (let funcId of globalRecordedAdditionalFunctionsMap.getOwnPropertyKeysArray()) {
+      let property = globalRecordedAdditionalFunctionsMap.properties.get(funcId);
+      invariant(property);
+      let funcValue = property.descriptor && property.descriptor.value;
+      invariant(funcValue instanceof FunctionValue);
```

Comment:
This can fail if diverged control-flow gets merged. Make it a nice user-facing error.